File: server/src/modules/rooms/domain/repositories/room.repository.ts
import {
  ParkingFees,
  RoomAddonAmenity,
  RoomEntity,
  RoomStatus,
} from '../entities/room.entity';

export type CreateRoomInput = {
  propertyId: string;
  title: string;
  area_sqm: number;
  monthly_rent: number;
  deposit_amount: number;
  electricity_rate_per_kwh: number;
  water_rate_per_m3: number;
  has_furniture?: boolean;
  description?: string;
  included_amenity_codes?: string[];
  addon_amenities?: RoomAddonAmenity[];
  images?: Array<{ url: string; sortOrder?: number }>;
  parking_fees?: ParkingFees;
};

export type UpdateRoomInput = {
  title?: string;
  status?: RoomStatus;
  area_sqm?: number;
  monthly_rent?: number;
  deposit_amount?: number;
  electricity_rate_per_kwh?: number;
  water_rate_per_m3?: number;
  has_furniture?: boolean;
  description?: string;
  included_amenity_codes?: string[];
  addon_amenities?: RoomAddonAmenity[];
  images?: Array<{ url: string; sortOrder?: number }>;
  parking_fees?: ParkingFees;
};

export abstract class RoomRepository {
  abstract createRoom(input: CreateRoomInput): Promise<RoomEntity>;
  abstract findAllByPropertyId(propertyId: string): Promise<RoomEntity[]>;
  abstract findById(id: string): Promise<RoomEntity | null>;
  abstract updateRoom(id: string, input: UpdateRoomInput): Promise<RoomEntity>;
  abstract deleteRoom(id: string): Promise<void>;
}

File: client/packages/domain/lib/src/room/repositories/room_repository.dart
import 'package:core/errors.dart';
import 'package:fpdart/fpdart.dart';

import '../../../room.dart';

abstract interface class RoomRepository {
  Future<Either<Failure, List<RoomEntity>>> getRooms({required String propertyId});
  Future<Either<Failure, RoomEntity>> getRoomById({required String id});
  Future<Either<Failure, RoomEntity>> createRoom({
    required String propertyId,
    required String title,
    required double areaSqm,
    required double monthlyRent,
    required double depositAmount,
    required double electricityRatePerKwh,
    required double waterRatePerM3,
    required List<String> includedAmenityCodes,
    required List<RoomAddonAmenity> addonAmenities,
    String? description,
    List<({String url, int sortOrder})>? images,
    RoomParkingFees? parkingFees,
  });
  Future<Either<Failure, RoomEntity>> updateRoom({
    required String id,
    String? title,
    RoomStatus? status,
    double? areaSqm,
    double? monthlyRent,
    double? depositAmount,
    double? electricityRatePerKwh,
    double? waterRatePerM3,
    List<String>? includedAmenityCodes,
    List<RoomAddonAmenity>? addonAmenities,
    String? description,
    List<({String url, int sortOrder})>? images,
    RoomParkingFees? parkingFees,
  });
  Future<Either<Failure, void>> deleteRoom({required String id});
}
